service/controller/internal/nodecontroller/controller.go

```diff
@@ -0,0 +1,106 @@
+package nodecontroller
+
+import (
+	"fmt"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/v4/pkg/controller"
+	"github.com/giantswarm/operatorkit/v4/pkg/resource"
+	"github.com/giantswarm/operatorkit/v4/pkg/resource/wrapper/retryresource"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/kvm-operator/pkg/label"
+	"github.com/giantswarm/kvm-operator/pkg/project"
+	"github.com/giantswarm/kvm-operator/service/controller/internal/nodecontroller/resource/podcondition"
+	"github.com/giantswarm/kvm-operator/service/controller/key"
+)
+
+type Config struct {
+	Cluster             v1alpha1.KVMConfig
+	ManagementK8sClient client.Client
+	WorkloadK8sClient   k8sclient.Interface
+	Logger              micrologger.Logger
+}
+
+type Controller struct {
+	*controller.Controller
+}
+
+func New(config Config) (*Controller, error) {
+	var err error
+
+	resources, err := newResources(config)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var operatorkitController *controller.Controller
+	{
+		c := controller.Config{
+			K8sClient: config.WorkloadK8sClient,
+			Logger:    config.Logger,
+			Resources: resources,
+			NewRuntimeObjectFunc: func() runtime.Object {
+				return new(corev1.Node)
+			},
+			Selector: labels.SelectorFromSet(map[string]string{
+				label.OperatorVersion: project.Version(),
+				"role":                key.WorkerID,
```

Comment:
I see Yulia had the same question in the time it took me to finish. I'd suggest adding a note there since we both noticed.
As I think about it, I guess we can't currently remove the master endpoint, but there is still the problem that workloads running on the master will continue to get routed traffic which will fail. I wonder if in the future we can either not rely solely on the Service endpoints ourselves, somehow drop only workload endpoints, or if this is just something we'll have to live with.

Reply:
I'll add a comment about why it's watching workers only.

I think as long as the master exists we should continue to route traffic to it even if it's not currently healthy. This might change once we move to multi-master on KVM as you say.